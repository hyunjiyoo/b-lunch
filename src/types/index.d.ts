export interface UserType {
  uid: string;
  displayName: string;
  email: string;
  isAdmin: boolean;
  photoURL: string;
}

export interface ProductType {
  name: string;
  price: number;
  category: string;
  description: string;
  option: string;
  file: FileList;  
}
