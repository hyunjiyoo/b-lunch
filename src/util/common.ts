import { initialUser } from "config/const";

const convertPriceFormat = (price: number) => {
  return 'â‚©' + (+price).toLocaleString('en');
}

const getUserInfo = () => {
  const localStorageUser = localStorage.getItem('user');
  const user = localStorageUser ? JSON.parse(localStorageUser) : initialUser;

  return user || initialUser;
}

const getCartItems = () => {
  return JSON.parse(localStorage.cart ?? JSON.stringify({}));
}

export { convertPriceFormat, getUserInfo, getCartItems };
