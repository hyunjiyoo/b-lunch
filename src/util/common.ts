import { initialUser } from 'config/const';

const convertPriceFormat = (price: number, won = '', unit = 'â‚©') => {
  return unit + (+price).toLocaleString('en') + won;
};

const getUserInfo = () => {
  const localStorageUser = localStorage.getItem('user');
  const user = localStorageUser ? JSON.parse(localStorageUser) : initialUser;

  return user || initialUser;
};

const getCartItems = () => {
  return JSON.parse(localStorage.cart ?? JSON.stringify({}));
};

export { convertPriceFormat, getUserInfo, getCartItems };
