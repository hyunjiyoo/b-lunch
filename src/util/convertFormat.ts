export const convertPriceFormat = (price: number) => {
  return 'â‚©' + (+price).toLocaleString('en');
}
