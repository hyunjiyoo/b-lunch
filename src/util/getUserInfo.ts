import { UserType } from 'types';

const initialUser: UserType = {
  uid: '',
  displayName: '',
  email: '',
  isAdmin: false,
  photoURL: '',
};

export const getUserFromLocalStorage = (): UserType => {
  const localStorageUser = localStorage.getItem('user');
  const user = localStorageUser ? JSON.parse(localStorageUser) : initialUser;
  return user;
};
