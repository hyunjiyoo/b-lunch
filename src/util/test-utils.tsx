import { ReactElement } from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

const customRender = (children: ReactElement) => {
  return render(<MemoryRouter>{children}</MemoryRouter>);
};

export * from '@testing-library/react';
export { customRender as render };
