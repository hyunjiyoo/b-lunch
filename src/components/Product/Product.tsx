import { useNavigate } from 'react-router-dom';
import { ProductType } from 'types';
import { convertPriceFormat } from 'util/common';

interface ProductProps {
  product: ProductType;
}

export default function Product({ product }: ProductProps) {
  const navigate = useNavigate();
  const { name, price, category, imageUrl } = product;

  const gotoDetailPage = () => {
    navigate(`/products/detail/${encodeURIComponent(product.id)}`, { state: { product }});
  }

  return (
    <li className='flex flex-col border border-slate-200 shadow-md cursor-pointer hover:opacity-95' onClick={gotoDetailPage}>
      <div className='h-220px bg-orange-50'>
        <img src={imageUrl} alt={`${name}_image`} className='w-180px min-w-180px p-2 h-full object-cover m-auto rounded hover:scale-y-105 ease-in duration-300' />
      </div>
      <div className='px-2 pt-1 flex justify-between min-w-fit'>
        <p className=''>{name}</p>
        <p className='text-right'>{convertPriceFormat(price)}</p>
      </div>
      <p className='mx-2 pb-1 text-zinc-400 text-sm flex-1'>{category}</p>
    </li>
  );
}
