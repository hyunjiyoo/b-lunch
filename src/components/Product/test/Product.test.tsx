import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { Route } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { ProductType } from 'types';
import { convertPriceFormat } from 'util/common';
import { withRouter } from 'tests/utils';
import Product from '../Product';

describe('Product', () => {
  const product: ProductType = {
    id: '1',
    name: 'Test Product',
    price: 1000,
    category: 'Test Category',
    description: 'Test description',
    option: 'Test Option',
    imageUrl: 'http://example.com/test.jpg',
  };

  it('renders product item', () => {
    render(withRouter(<Route path='/' element={<Product product={product} />} />));

    const image = screen.getByRole('img') as HTMLImageElement;
    expect(image.alt).toBe(`${product.name}_image`);
    expect(image.src).toBe(product.imageUrl);

    expect(screen.getByText(product.name)).toBeInTheDocument();
    expect(screen.getByText(convertPriceFormat(product.price))).toBeInTheDocument();
    expect(screen.getByText(product.category)).toBeInTheDocument();
  });

  it('renders correctly', () => {
    const component = renderer.create(withRouter(<Route path='/' element={<Product product={product} />} />));
    expect(component.toJSON()).toMatchSnapshot();
  });
});
