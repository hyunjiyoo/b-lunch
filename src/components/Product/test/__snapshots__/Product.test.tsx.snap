// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product renders correctly 1`] = `
<li
  className="flex flex-col border border-slate-200 shadow-md cursor-pointer hover:opacity-95"
  onClick={[Function]}
>
  <div
    className="h-220px bg-orange-50"
  >
    <img
      alt="Test Product_image"
      className="w-180px min-w-180px p-2 h-full object-cover m-auto rounded hover:scale-y-105 ease-in duration-300"
      src="http://example.com/test.jpg"
    />
  </div>
  <div
    className="px-2 pt-1 flex justify-between min-w-fit"
  >
    <p
      className=""
    >
      Test Product
    </p>
    <p
      className="text-right"
    >
      â‚©1,000
    </p>
  </div>
  <p
    className="mx-2 pb-1 text-zinc-400 text-sm flex-1"
  >
    Test Category
  </p>
</li>
`;
